safekeeper/src/wal_storage.rs

```diff
@@ -223,15 +215,33 @@ impl PhysicalStorage {
             Ok((file, true))
         } else {
             // Create and fill new partial file
+            //
+            // We're using fdatasync during WAL writing, so file size must not
+            // change; to this end it is filled with zeros here. To avoid using
+            // half initialized segment, first bake it under tmp filename and
+            // then rename.
+            let tmp_path = self.timeline_dir.join("waltmp");
             let mut file = OpenOptions::new()
                 .create(true)
                 .write(true)
-                .open(&wal_file_partial_path)
+                .open(&tmp_path)
                 .await
-                .with_context(|| format!("Failed to open log file {:?}", &wal_file_path))?;
+                .with_context(|| format!("Failed to open tmp wal file {:?}", &tmp_path))?;
 
             write_zeroes(&mut file, self.wal_seg_size).await?;
-            self.fsync_file(&file).await?;
+
+            // Note: this doesn't get into observe_flush_seconds metric. But
+            // segment init should be separate metric, if any.
+            if let Err(e) =
+                durable_rename(&tmp_path, &wal_file_partial_path, !self.conf.no_sync).await
```

Comment:
We need to check if extra fsync affect performance. I think most of the time is spent in `write_zeroes` so `rename+fsync` shouldn't be noticeable, but better check.

Reply:
Tried to check it locally with 
https://github.com/neondatabase/neon/tree/sk-basic-bench
```
rm -rf pg_install/ && BUILD_TYPE=release CARGO_BUILD_FLAGS="--features=testing" make -s -j12
export DEFAULT_PG_VERSION=16; export NEON_BIN=./target/release; export TEST_OUTPUT=/nenc/test_output
pkill pageserver; pkill postgres; pkill safekeeper; pkill pytest; pkill compute_ctl; rmc; rm -rf $TEST_OUTPUT/ && pytest test_runner/performance  --capture=no -k "test_pgbench[" --preserve-database-files
```

This is a basic pgbench init test without waiting for pageserver (disable backpressure, don't wait for last_record_lsn and ps checkpoint). It is still far from being optimal for benchmarking safekeepers (pageserver is not fully excluded, can run only locally), but ok as a baseline, I think we should commit something like this.

This PR doesn't have visible impact locally on its results -- about 131 Mib/s on my laptop, dividing sk data dir size by test duration.